src/tools/info/proto_info.c

```diff
@@ -134,32 +134,155 @@ ucs_status_t wait_completion(ucp_worker_h worker, ucs_status_ptr_t status_ptr)
+    if (status != UCS_OK) {
+        printf("<Failed to get port>\n");
+        goto out_destroy_listener;
+    }
+
+    *listener_p = listener;
+out:
+    return status;
+
+out_destroy_listener:
+    ucp_listener_destroy(listener);
+    goto out;
+}
+
+ucs_status_t
+print_ucp_ep_info(ucp_worker_h worker, const ucp_ep_params_t *base_ep_params,
+                  const char *ip_addr)
+{
+    ucp_listener_h listener = NULL;
+    ucp_ep_h server_ep      = NULL;
+    ucs_status_t status, tmp_status;
+    ucs_status_ptr_t status_ptr;
+    ucp_ep_params_t ep_params;
+    ucp_address_t *address;
+    size_t address_length;
+    struct sockaddr_in connect_saddr;
+    uint16_t listen_port;
+    ucp_ep_h ep;
+    ucp_request_param_t request_param;
+
+    ep_params = *base_ep_params;
+
+    if (strlen(ip_addr) != 0) {
+        status = create_listener(worker, &listener, &listen_port, &server_ep);
+        if (status != UCS_OK) {
+            return status;
+        }
+
+        set_connect_addr(ip_addr, listen_port, &connect_saddr);
+
+        ep_params.field_mask      |= UCP_EP_PARAM_FIELD_FLAGS |
+                                     UCP_EP_PARAM_FIELD_SOCK_ADDR;
+        ep_params.flags            = UCP_EP_PARAMS_FLAGS_CLIENT_SERVER;
+        ep_params.sockaddr.addr    = (struct sockaddr*)&connect_saddr;
+        ep_params.sockaddr.addrlen = sizeof(connect_saddr);
+    } else {
+        status = ucp_worker_get_address(worker, &address, &address_length);
+        if (status != UCS_OK) {
+            printf("<Failed to get UCP worker address>\n");
+            return status;
+        }
+
+        ep_params.field_mask |= UCP_EP_PARAM_FIELD_REMOTE_ADDRESS;
+        ep_params.address     = address;
+    }
+
+    status = ucp_ep_create(worker, &ep_params, &ep);
+    if (strlen(ip_addr) == 0) {
+        ucp_worker_release_address(worker, address);
+    }
+    if (status != UCS_OK) {
+        printf("<Failed to create UCP endpoint>\n");
+        goto out_destroy_listener;
+    }
+
+    request_param.op_attr_mask = 0;
+    /* do EP flush to make sure that fully completed to a peer and final
+     * configuration is applied */
+    status_ptr = ucp_ep_flush_nbx(ep, &request_param);
+    status     = wait_completion(worker, status_ptr);
+    if (status != UCS_OK) {
+        printf("<Failed to flush UCP endpoint>\n");
+    }
+
+    ucp_ep_print_info(ep, stdout);
+
+    status_ptr = ucp_disconnect_nb(ep);
+    tmp_status = wait_completion(worker, status_ptr);
+    if ((tmp_status != UCS_OK) && (status == UCS_OK)) {
+        printf("<Failed to destroy UCP endpoint>\n");
+        status = tmp_status;
+    }
+
+    if (server_ep != NULL) {
+        status_ptr = ucp_ep_close_nb(server_ep, UCP_EP_CLOSE_MODE_FORCE);
+        tmp_status = wait_completion(worker, status_ptr);
+        if ((tmp_status != UCS_OK) && (status == UCS_OK)) {
+            printf("<Failed to destroy server UCP endpoint>\n");
+            status = tmp_status;
+        }
+    }
```

Comment:
pls have a common helper func to blocking-close a ucp endpoint

Reply:
done